_posts/2018-08-01-toykio.md

```diff
@@ -0,0 +1,198 @@
+---
+layout: post
+title:  "Toykio"
+author: "Alexander Polakov"
+author_github: "polachok"
+date:   2018-08-01
+categories: blog
+---
+
+What is toykio?
+---------------
+[toykio](https://github.com/polachok/toykio) is a simple futures executor intended for learning about
+how executors work. Toykio only provides minimal features: event loop, TcpStream and TcpListener, but it 
+turns out that due to futures being composable this is enough to build complex clients and servers.
+
+I'll give a quick overview of toykio components.
+
+TcpStream
+---------
+AsyncTcpStream is defined as a newtype wrapper around TcpStream from standard library. `connect`
+function opens a connection, just like in std, and sets socket to nonblocking mode. This means
+that read() and write() functions will return immediately, either with data or an error. If there's not enough data (for reads) or buffer space (for writes) avaialble, a special kind of error - `WouldBlock` - is returned. We'll discuss how to handle them in the next section.
+
+AsyncRead & AsyncWrite
+----------------------
+AsyncRead and AsyncWrite traits are the foundation of all I/O futures. FutureReadExt and FutureWriteExt
+extension methods (like `write_all`) are built on top of them. This makes `futures` crate independent of
+the event loop implementation.
+`AsyncRead` and `AsyncWrite` implementations provide this connection between futures and the reactor. 
+
+Let's take a look at 
+[AsyncRead](https://github.com/polachok/toykio/blob/futures-0.3/src/lib.rs#L327) for example:
+
+```rust
+impl AsyncRead for AsyncTcpStream {
+    fn poll_read(&mut self, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, Error>> {
+	// get TcpStream file descriptor
+        let fd = self.0.as_raw_fd();
+        let waker = cx.waker();
+
+        match self.0.read(buf) {
+            Ok(len) => Poll::Ready(Ok(len)),
+            Err(ref err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                REACTOR.with(|reactor| reactor.add_read_interest(fd, waker.clone()));
+
+                Poll::Pending
+            }
+            Err(err) => panic!("error {:?}", err),
+        }
+    }
+}
+```
+It tries to perform a read on the underlying TcpStream, and registers its interest in `readable`
+event if `WouldBlock` is returned. This is explained in the next section.
+
+[AsyncWrite](https://github.com/polachok/toykio/blob/futures-0.3/src/lib.rs#L346) does
```

Comment:
The `AsyncWrite` implementation for `AsyncTcpStream` does something similar for writes.